concurrency/StandardMatrixRequestGenerator.java

```diff
@@ -0,0 +1,123 @@
+    private final Queue<PostRevisionRequest> postRevisionRequests = new ConcurrentLinkedQueue<>();
+    private final Queue<PutMatrixRequest> putMatrixRequests = new ConcurrentLinkedQueue<>();
+    private final Queue<DeleteMatrixRequest> deleteMatrixRequests = new ConcurrentLinkedQueue<>();
+    private final Queue<DiffRevisionRequest> diffRevisionRequests = new ConcurrentLinkedQueue<>();
+    
+    public StandardMatrixRequestGenerator withGetMatrixRequest(GetMatrixRequest... req)
+    {
+        getMatrixRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }
+    
+    public StandardMatrixRequestGenerator withGetRevisionRequest(GetRevisionRequest... req)
+    {
+        getRevisionRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }    
+    public StandardMatrixRequestGenerator withPutMatrixRequest(PutMatrixRequest... req)
+    {
+        putMatrixRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }    
+    public StandardMatrixRequestGenerator withPostRevisionRequest(PostRevisionRequest... req)
+    {
+        postRevisionRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }    
+    public StandardMatrixRequestGenerator withDiffRevisionRequest(DiffRevisionRequest... req)
+    {
+        diffRevisionRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }    
+    public StandardMatrixRequestGenerator withDeleteMatrixRequest(DeleteMatrixRequest... req)
+    {
+        deleteMatrixRequests.addAll(Arrays.stream(req).collect(Collectors.toList()));
+        return this;
+    }
+    
+    @Override
+    public GetMatrixRequest getNextGetMatrixRequest(String matrixId)
+    {
+        GetMatrixRequest req = getMatrixRequests.peek();
+        if (Objects.isNull(req)) throw new RuntimeException("No GetMatrixRequest provided during construction.");
+        if (getMatrixRequests.size()==1) return req;
+        return getMatrixRequests.poll();
+    }
+
+    @Override
+    public GetRevisionRequest getNextGetRevisionRequest(String matrixId, String basisRevisionId)
+    {
+        GetRevisionRequest req = getRevisionRequests.peek();
+        if (Objects.isNull(req)) throw new RuntimeException("No GetRevisionRequest provided during construction.");
+        if (getRevisionRequests.size()==1) return req;
+        return getRevisionRequests.poll();
+    }
+
+    @Override
+    public PostRevisionRequest getNextPostRevisionRequest(String matrixId, String basisRevisionId, Revision revisionToPost)
+    {
+        PostRevisionRequest req = postRevisionRequests.peek();
+        if (Objects.isNull(req)) throw new RuntimeException("No PostRevisionRequest provided during construction.");
+        if (postRevisionRequests.size()==1) return req;
+        return postRevisionRequests.poll();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentLinkedQueue`, but your usage of `size()` and `poll()` may not be thread-safe at lines: **87 and 88**. If the `ConcurrentLinkedQueue` contains only **one element** and two threads perform this **same check at the same time** , the second thread&#39;s call to `poll()` will get `null`. This `null` may be used at line: **88**.

**Fix**  
 Consider calling `poll()`, checking instead of your current check if the returned object is `null`, and only then using that object.